### PR TITLE
feat: Add Oracle extractor loader to Amundsen

### PR DESCRIPTION
**Pull request by savioteles**
Signed-off-by: Savio <savioteles@gmail.com>

### Summary of Changes

This PR adds a new Oracle extractor to load the database metadata into Amundsen including a sample example.

### Tests

Added a new unit test for Oracle extractor.

### CODE REVIEW

File: databuilder/databuilder/extractor/oracle_metadata_extractor.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    def get_sql_statement(self, use_catalog_as_cluster_name: bool, where_clause_suffix: str) -> str:
+        if use_catalog_as_cluster_name:
+            cluster_source = "c.table_catalog"
+        else:
+            cluster_source = f"'{self._cluster}'"
+
+        return """
+        SELECT 
+            {cluster_source} as "cluster",
+            lower(c.OWNER) as "schema",
```
Comment (mgorsk1) [inline databuilder/databuilder/extractor/oracle_metadata_extractor.py:27]: can you unify all table & column names to either lower/upper case? I think oracle doesn't have case-sensitive column and table names so this sql could be cleaner

File: databuilder/databuilder/extractor/base_oracle_metadata_extractor.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    DEFAULT_CLUSTER_NAME = 'master'
+
+    DEFAULT_CONFIG = ConfigFactory.from_dict(
+        {WHERE_CLAUSE_SUFFIX_KEY: ' ', CLUSTER_KEY: DEFAULT_CLUSTER_NAME, USE_CATALOG_AS_CLUSTER_NAME: True}
+    )
+
+    @abc.abstractmethod
+    def get_sql_statement(self, use_catalog_as_cluster_name: bool, where_clause_suffix: str) -> Any:
+        """
+        :return: Provides a record or None if no more to extract
```
Comment (mgorsk1) [inline databuilder/databuilder/extractor/base_oracle_metadata_extractor.py:45]: is this accurate description of this function?

File: databuilder/databuilder/extractor/base_oracle_metadata_extractor.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        self.sql_stmt = self.get_sql_statement(
+            use_catalog_as_cluster_name=conf.get_bool(BaseOracleMetadataExtractor.USE_CATALOG_AS_CLUSTER_NAME),
+            where_clause_suffix=conf.get_string(BaseOracleMetadataExtractor.WHERE_CLAUSE_SUFFIX_KEY),
+        )
+
+        self._alchemy_extractor = SQLAlchemyExtractor()
+        sql_alch_conf = Scoped.get_scoped_conf(conf, self._alchemy_extractor.get_scope())\
+            .with_fallback(ConfigFactory.from_dict({SQLAlchemyExtractor.EXTRACT_SQL: self.sql_stmt}))
+
+        self.sql_stmt = sql_alch_conf.get_string(SQLAlchemyExtractor.EXTRACT_SQL)
```
Comment (mgorsk1) [inline databuilder/databuilder/extractor/base_oracle_metadata_extractor.py:64]: why do we need this if it's already defined in line 55?

File: databuilder/databuilder/extractor/base_oracle_metadata_extractor.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+            for row in group:
+                last_row = row
+                columns.append(ColumnMetadata(row['col_name'], row['col_description'],
+                                              row['col_type'], row['col_sort_order']))
+
+            yield TableMetadata(self._database, last_row['cluster'],
+                                last_row['schema'],
+                                last_row['name'],
+                                last_row['description'],
```
Comment (mgorsk1) [inline databuilder/databuilder/extractor/base_oracle_metadata_extractor.py:95]: how will it behave if description is null (which is possible when using left join from OracleMetadataExtractor)? Maybe it's safer to collect every property with `.get('property', '')`

File: databuilder/databuilder/extractor/base_oracle_metadata_extractor.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+class BaseOracleMetadataExtractor(Extractor):
+    """
+    Extracts Oracle table and column metadata from underlying meta store database using SQLAlchemyExtractor
+    """
+
+    # CONFIG KEYS
+    WHERE_CLAUSE_SUFFIX_KEY = 'where_clause_suffix'
+    CLUSTER_KEY = 'cluster_key'
+    USE_CATALOG_AS_CLUSTER_NAME = 'use_catalog_as_cluster_name'
+    DATABASE_KEY = 'database_key'
```
Comment (mgorsk1) [inline databuilder/databuilder/extractor/base_oracle_metadata_extractor.py:33]: shouldn't this default to `oracle`?

**Review by mgorsk1 — COMMENTED** [review]
few nits/questions

one additional suggestion - you could also try adding oracle icon to frontend so it's distinguished nicely in UI. As an example you can take a look at this PR (focus only on changes in `frontend` catalog): https://github.com/amundsen-io/amundsen/pull/1100

File: databuilder/databuilder/extractor/oracle_metadata_extractor.py (hunk 1 above)
Comment (savioteles) [reply databuilder/databuilder/extractor/oracle_metadata_extractor.py]: Tks. Done!

File: databuilder/databuilder/extractor/base_oracle_metadata_extractor.py (hunk 2 above)
Comment (savioteles) [reply databuilder/databuilder/extractor/base_oracle_metadata_extractor.py:45]: OracleMetadataExtractor class overrides this function returning a SQL. When the function does not exist, it returns null.

File: databuilder/databuilder/extractor/base_oracle_metadata_extractor.py (hunk 4 above)
Comment (savioteles) [reply databuilder/databuilder/extractor/base_oracle_metadata_extractor.py:95]: The description field is not required in TableMetadata class. CassandraExtractor class is an example:

File: databuilder/databuilder/extractor/base_oracle_metadata_extractor.py (hunk 5 above)
Comment (savioteles) [reply databuilder/databuilder/extractor/base_oracle_metadata_extractor.py:33]: It's done on line 53.

File: databuilder/databuilder/extractor/base_oracle_metadata_extractor.py (hunk 2 above)
Comment (mgorsk1) [reply databuilder/databuilder/extractor/base_oracle_metadata_extractor.py:45]: yeah but does the function really do what this description says?

**Comment by savioteles** [comment]
@mgorsk1, I have added oracle icon to frontend in the last commit.

File: databuilder/databuilder/extractor/base_oracle_metadata_extractor.py (hunk 2 above)
Comment (savioteles) [reply databuilder/databuilder/extractor/base_oracle_metadata_extractor.py]: That's true! Description updated.

File: databuilder/databuilder/extractor/base_oracle_metadata_extractor.py (hunk 3 above)
Comment (savioteles) [reply databuilder/databuilder/extractor/base_oracle_metadata_extractor.py:64]: It's just the fallback. Look at the line 61

File: databuilder/setup.py [hunk 6]
```diff
@@ -76,7 +76,8 @@
 
 rds = [
     'sqlalchemy>=1.3.6,<1.4',
-    'mysqlclient>=1.3.6,<3'
+    'mysqlclient>=1.3.6,<3',
+    'cx_Oracle==8.2.1'
```
Comment (mgorsk1) [inline databuilder/setup.py:80]: I think rds is meant for something else, not collection of libs regarding all rds systems. you can just create separate oracle group

File: databuilder/databuilder/extractor/oracle_metadata_extractor.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+from databuilder.extractor.base_oracle_metadata_extractor import BaseOracleMetadataExtractor
+
+
+class OracleMetadataExtractor(BaseOracleMetadataExtractor):
+    """
+    Extracts Oracle table and column metadata from underlying meta store database using SQLAlchemyExtractor
+    """
+
+    def get_sql_statement(self, use_catalog_as_cluster_name: bool, where_clause_suffix: str) -> str:
+        if use_catalog_as_cluster_name:
```
Comment (mgorsk1) [inline databuilder/databuilder/extractor/oracle_metadata_extractor.py:19]: This feels unnecessary. Why would we use table name as cluster name? I'd keep it more simple and just have default cluster name + get rid of this book completely.

**Review by mgorsk1 — CHANGES_REQUESTED** [review]

File: databuilder/setup.py (hunk 6 above)
Comment (savioteles) [reply databuilder/setup.py]: Done!

File: databuilder/databuilder/extractor/oracle_metadata_extractor.py (hunk 7 above)
Comment (savioteles) [reply databuilder/databuilder/extractor/oracle_metadata_extractor.py]: You're right. I've ceated a default cluster name to Oracle extractor in line 51 (file base_oracle_metadata_extractor.py).

File: databuilder/databuilder/extractor/oracle_metadata_extractor.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+from typing import (  # noqa: F401
+    Any, Dict, Iterator, Union,
+)
+
+from pyhocon import ConfigFactory, ConfigTree  # noqa: F401
+
+from databuilder.extractor.base_oracle_metadata_extractor import BaseOracleMetadataExtractor
+
+
+class OracleMetadataExtractor(BaseOracleMetadataExtractor):
```
Comment (sewardgw) [inline databuilder/databuilder/extractor/oracle_metadata_extractor.py:13]: Why is there a `BaseOracleMetadataExtractor` as well as an `OracleMetadataExtractor`? This approach makes sense for generic SqlAlchemy extractor which can theoretically be extended to any type of relational database as well as for Postgres which can serve both Postgres and Redshift. Are there multiple implementations of Oracle that are slightly different which may require separate implementations?

File: databuilder/example/scripts/sample_oracle_loader.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+
+# todo: connection string needs to change
+def connection_string():
+    user = 'username'
+    password = 'password'
+    host = 'localhost'
+    port = '1521'
+    service = 'service'
+    return "oracle+cx_oracle://%s:%s@%s:%s/%s" % (user, password, host, port, service)
```
Comment (sewardgw) [inline databuilder/example/scripts/sample_oracle_loader.py:64]: Does this require any specific Oracle drivers to be on the path to run or are all of the drivers included in the distribution of the `cx_oracle` package?

File: databuilder/setup.py [hunk 10]
```diff
@@ -74,13 +74,17 @@
     'apache-atlas>=0.0.11'
 ]
 
+oracle = [
+    'cx_Oracle==8.2.1'
```
Comment (sewardgw) [inline databuilder/setup.py:78]: Does this require any specific Oracle drivers to be on the path to run or are all of the drivers included in the distribution of the `cx_oracle` package?

File: databuilder/databuilder/extractor/oracle_metadata_extractor.py (hunk 8 above)
Comment (savioteles) [reply databuilder/databuilder/extractor/oracle_metadata_extractor.py]: You're right! I merged the two classes into one.

File: databuilder/example/scripts/sample_oracle_loader.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+if len(sys.argv) > 2:
+    neo_host = sys.argv[2]
+
+es = Elasticsearch([
+    {'host': es_host if es_host else 'localhost'},
+])
+
+DB_FILE = '/tmp/test.db'
+SQLITE_CONN_STRING = 'sqlite:////tmp/test.db'
+Base = declarative_base()
```
Comment (mgorsk1) [inline databuilder/example/scripts/sample_oracle_loader.py:45]: Do we use this anywhwere (and two rows above)?

File: databuilder/README.md [hunk 12]
```diff
@@ -253,6 +253,26 @@ job = DefaultJob(
 job.launch()
 ```
 
+#### [OracleMetadataExtractor](https://github.com/amundsen-io/amundsen/blob/main/databuilder/databuilder/extractor/oracle_metadata_extractor.py "OracleMetadataExtractor")
+An extractor that extracts table and column metadata including database, schema, table name, table description, column name and column description from the Oracle database.
+
+By default, the Oracle database name is used as the cluster name. To override this, set `USE_CATALOG_AS_CLUSTER_NAME` to `False`, and `CLUSTER_KEY` to what you wish to use as the cluster name.
```
Comment (mgorsk1) [inline databuilder/README.md:259]: Is it still true?

File: databuilder/README.md [hunk 13]
```diff
@@ -253,6 +253,26 @@ job = DefaultJob(
...
+
+The `where_clause_suffix` below should define which schemas you'd like to query. The SQL query driving the extraction is defined [here](https://github.com/amundsen-io/amundsen/blob/main/databuilder/databuilder/extractor/oracle_metadata_extractor.py)
+
+```python
+job_config = ConfigFactory.from_dict({
+    'extractor.oracle_metadata.{}'.format(OracleMetadataExtractor.WHERE_CLAUSE_SUFFIX_KEY): where_clause_suffix,
+    'extractor.oracle_metadata.{}'.format(OracleMetadataExtractor.USE_CATALOG_AS_CLUSTER_NAME): True,
```
Comment (mgorsk1) [inline databuilder/README.md:266]: Is it still valid?

File: databuilder/databuilder/extractor/oracle_metadata_extractor.py [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+class OracleMetadataExtractor():
+    """
+    Extracts Oracle table and column metadata from underlying meta store database using SQLAlchemyExtractor
+    """
+
+    # CONFIG KEYS
+    WHERE_CLAUSE_SUFFIX_KEY = 'where_clause_suffix'
+    CLUSTER_KEY = 'cluster_key'
+    USE_CATALOG_AS_CLUSTER_NAME = 'use_catalog_as_cluster_name'
```
Comment (mgorsk1) [inline databuilder/databuilder/extractor/oracle_metadata_extractor.py:30]: Is this still used ?

File: databuilder/setup.py (hunk 10 above)
Comment (savioteles) [reply databuilder/setup.py:78]: Yes. The cx_Oracle module loads Oracle Client libraries which communicate over Oracle Net to an existing database.

File: databuilder/example/scripts/sample_oracle_loader.py (hunk 11 above)
Comment (savioteles) [reply databuilder/example/scripts/sample_oracle_loader.py]: No. These lines have been removed

File: databuilder/README.md (hunk 12 above)
Comment (savioteles) [reply databuilder/README.md]: No. Fixed.

File: databuilder/README.md (hunk 13 above)
Comment (savioteles) [reply databuilder/README.md]: No. I deleted this line.

File: databuilder/databuilder/extractor/oracle_metadata_extractor.py (hunk 14 above)
Comment (savioteles) [reply databuilder/databuilder/extractor/oracle_metadata_extractor.py]: No. I deleted this line.

**Review by mgorsk1 — APPROVED** [review]
thanks, awesome stuff!

**Review by mgorsk1 — APPROVED** [review]

**Comment by mgorsk1** [comment]
some unit tests failing in front end @savioteles - would you mind taking a look?

**Comment by savioteles** [comment]
@mgorsk1, unit tests fixed!

**Comment by mgorsk1** [comment]
still some linting with databuilder @savioteles - you can check if it will pass checks locally by running `make test` inside `databuilder` folder

**Comment by savioteles** [comment]
@mgorsk1, databuilder unit tests fixed. I have also fixed the front end unit tests (oracle's icon path was wrong)

**Comment by mgorsk1** [comment]
thanks @savioteles !

**Pull request merged**